domain/consensus/ruleerrors/rule_error.go

```diff
@@ -232,13 +234,43 @@ var (
 // specifically due to a rule violation.
 type RuleError struct {
 	message string
+	inner   error
 }
 
 // Error satisfies the error interface and prints human-readable errors.
 func (e RuleError) Error() string {
+	if e.inner != nil {
+		return e.message + ": " + e.inner.Error()
+	}
 	return e.message
 }
 
+// Unwrap satisfies the errors.Unwrap interface
+func (e RuleError) Unwrap() error {
+	return e.inner
+}
+
+// Cause satisfies the github.com/pkg/errors.Cause interface
+func (e RuleError) Cause() error {
+	return e.inner
+}
+
 func newRuleError(message string) RuleError {
-	return RuleError{message: message}
+	return RuleError{message: message, inner: nil}
+}
+
+// ErrMissingTxOut indicates a transaction output referenced by an input
+// either does not exist or has already been spent.
+type ErrMissingTxOut struct {
+	MissingOutpoints []externalapi.DomainOutpoint
+}
+
+func (e ErrMissingTxOut) Error() string {
+	return fmt.Sprint(e.MissingOutpoints)
+}
+
+// WrapInRuleError wraps any error inside a RuleError
+func WrapInRuleError(err error) error {
```

Comment:
Instead of externalizing WrapInRuleError, I would've preferred a function called NewErrMissingTxOut that creates `ErrMissingTxOut`, wraps it in a `RuleError`, and returns it as an `error`.

Reply:
And then we will have multiple of those per error? (on top of the `Error()` impls)

Reply:
You mean for other errors? If so, then yes. I think I prefer that over the possibility of someone accidentally returning a not-wrapped ErrXxx.

Reply:
Done